src/sage/stats/distributions/discrete_gaussian_lattice.py

```diff
@@ -222,51 +229,214 @@ def _normalisation_factor_zz(self, tau=3):
+            BOUND = max(1, (self._RR(log(10**4, self.n)).ceil() - 1) // 2)
+            if BOUND > 10:
+                BOUND = 10
+            coords = itertools.product(range(-BOUND, BOUND + 1), repeat=self.n)
+            return sum(self.f((vector(u) + base) * self.B) for u in coords)
+
+        if self.B.nrows() != self.B.ncols():
+            raise NotImplementedError("Basis must be a square matrix for now.")
+
+        if self.is_spherical and not self._c_in_lattice:
+            raise NotImplementedError("Lattice must contain 0 for now.")
+
+        if self.B.base_ring() != ZZ:
+            raise NotImplementedError("Lattice must be integral for now.")
+
         sigma = self._sigma
-        return sum(f(x) for x in _iter_vectors(n, -ceil(tau * sigma),
-                                               ceil(tau * sigma)))
+        prec = DiscreteGaussianDistributionLatticeSampler.compute_precision(
+            prec, sigma
+        )
+        R = RealField(prec=prec)
+        if sigma > 1:
+            det = self.B.det()
+            norm_factor = (sigma * sqrt(2 * pi))**self.n / det
+        else:
+            det = 1
+            norm_factor = 1
+
+        # qfrep computes theta series of a quadratic form, which is *half* the
+        # generating function of number of vectors with given norm (and no 0)
+        Q = self.Q
+        if tau is not None:
+            freq = Q.__pari__().qfrep(tau * sigma, 0)
+            res = R(1)
+            for x, fq in enumerate(freq):
+                res += 2 * ZZ(fq) * f_or_hat((x + 1) / det**self.n)
+            return R(norm_factor * res)
+
+        res = R(1)
+        bound = 0
+        # There might still be precision issue but whatever
+        while True:
+            bound += 1
+            cnt = ZZ(Q.__pari__().qfrep(bound, 0)[bound - 1])
+            inc = 2 * cnt * f_or_hat(bound / det**self.n)
+            if cnt > 0 and res == res + inc:
+                return R(norm_factor * res)
+            res += inc
+
+    @cached_method
+    def _maximal_r(self):
+        r"""
+        This function computes the largest value `r > 0` such that `Σ - r²BBᵀ`
+        is positive definite.
+
+        This is equivalent to finding `λ₁(Σ / Q) = 1 / λₙ(Q / Σ)`, which is done
+        via the Power iteration method.
+
+        EXAMPLES::
+
+            sage: from sage.stats.distributions.discrete_gaussian_lattice import DiscreteGaussianDistributionLatticeSampler as DGL
+            sage: n = 3
+            sage: Sigma = Matrix(ZZ, [[5, -2, 4], [-2, 10, -5], [4, -5, 5]])
+            sage: c = vector(ZZ, [7, 2, 5])
+            sage: D = DGL(ZZ^n, Sigma, c)
+            sage: r = D._maximal_r(); r
+            0.58402...
+            sage: e_vals = (D.sigma() - r^2 * D.Q).eigenvalues()
+            sage: assert all(e_val >= -1e-12 for e_val in e_vals)
+        """
+        if self.is_spherical:
+            raise RuntimeError("You have encountered a bug. File it! :)")
```

Comment:
```suggestion
        assert self.is_sperical
```
I am not `00% sure versus the `RuntimeError. I think the assert is slightly better here, but I am not sure I could explain why I feel that way.